### PR TITLE
nixos-generate-config: enable overriding configuration.nix

### PR DESCRIPTION
**Pull request by basvandijk**
I'm writing a custom NixOS installation ISO image for installing various machines. The image contains an interactive script that performs an automatic installation. After asking which disk to install to it will partition the disk, format the filesystems, mount them and then it will call `nixos-generate-config` to generate the initial `/etc/nixos/configuration.nix`. I would like to automatically set some things in this initial configuration like enabling SSH, setting `users.users.root.openssh.authorizedKeys.keyFiles` and maybe some more settings. I could of course patch or overwrite `/etc/nixos/configuration.nix` but that could destroy potentially useful stuff like the `$bootLoaderConfig`. Overriding it in NixOS would be safer.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edolstra @grahamc

### CODE REVIEW

File: nixos/modules/installer/tools/tools.nix [hunk 1]
```diff
@@ -61,8 +64,111 @@ in
...
+        # users.users.jane = {
+        #   isNormalUser = true;
+        #   extraGroups = [ "wheel" ]; # Enable ‘sudo’ for the user.
+        # };
+
+        # This value determines the NixOS release with which your system is to be
+        # compatible, in order to avoid breaking some software such as database
+        # servers. You should change this only after NixOS release notes say you
+        # should.
+        system.stateVersion = "''${\(qw(@release@))}"; # Did you read the comment?
```
Comment (grahamc) [inline nixos/modules/installer/tools/tools.nix]: Hrm, this is pretty ugly. Can the substitution be done here, and not in a perl escape?

**Comment by basvandijk** [comment]
@GrahamcOfBorg test nixos-generate-config

File: nixos/modules/installer/tools/tools.nix (hunk 1 above)
Comment (basvandijk) [reply nixos/modules/installer/tools/tools.nix]: Great point. That should be possible I think. I'll fix it.

Comment (edolstra) [reply nixos/modules/installer/tools/tools.nix]: Actually that seems worse, because then we have *two* kinds of substitution going on in this file.

**Comment by edolstra** [comment]
Hm, the problem with `system.nixos-generate-config.configuration` is that it's very tightly coupled with the implementation details of `nixos-generate-config`. Thus it exposes variable names like `$bootLoaderConfig` and requires `nixos-generate-config` to be written in Perl. So the existence of this option would make future changes to `nixos-generate-config` much harder.

**Comment by basvandijk** [comment]
>  So the existence of this option would make future changes to nixos-generate-config much harder.

That's true. I did make `system.nixos-generate-config.configuration` an `internal` option so that it doesn't show up in the documentation. 

I could add to the option description: 

> We don't guarantee any backward compatibly. Use at your own risk!

File: nixos/modules/installer/tools/tools.nix (hunk 1 above)
Comment (basvandijk) [reply nixos/modules/installer/tools/tools.nix]: Well in my original commit there were actually already two substitutions going on. Now it's back to a single `substituteAll` call as before.

**Comment by basvandijk** [comment]
@GrahamcOfBorg test nixos-generate-config

**Comment by edolstra** [comment]
@basvandijk Ah, I didn't notice you made it internal. In that case it's fine.

**Pull request merged**